luatest/server.lua

```diff
@@ -547,24 +538,44 @@ function Server:exec(fn, args, options)
         error('net_box is not connected', 2)
     end
 
-    local ups = utils.upvalues(fn)
-    if next(ups) ~= nil then
+    local autorequired_pkgs = {'luatest'}
+    local autorequired_ups = {}
+    local other_ups = {}
+    for i = 1, debug.getinfo(fn, 'u').nups do
+        local is_other_up = true
+        local name, value = debug.getupvalue(fn, i)
+        for _, p in ipairs(autorequired_pkgs) do
+            if package.loaded[p] == value then
+                autorequired_ups[i] = p
+                is_other_up = false
+                break
+            end
+        end
+        if is_other_up then
+            table.insert(other_ups, name)
+        end
+    end
+
+    if next(other_ups) ~= nil then
         local err = string.format(
             'bad argument #2 to exec (excess upvalues: %s)',
-            table.concat(ups, ', ')
+            table.concat(other_ups, ', ')
         )
         error(err, 2)
     end
 
-    return exec_tail(self:eval([[
-        local dump, args = ...
+    return exec_tail(self.net_box:eval([[
+        local dump, args, autorequired_ups = ...
         local fn = loadstring(dump)
+        for i, v in pairs(autorequired_ups) do
+            debug.setupvalue(fn, i, require(v))
+        end
```

Comment:
I'm not sure that upvalues numeration is guaranteed to be the same for the original function and its `loadstring()` of `dump()`. I would rather loop over existing upvalues and match it with a module name by the upvalue name.

Reply:
Fixed.